src/gui/editors/AutomationEditor.cpp

```diff
@@ -2238,7 +1903,58 @@ void AutomationEditor::updateTopBottomLevels()
 
 
 
+/* @brief Given a mouse coordinate, returns a timeMap::iterator that points to
+ *        the latest node that is inside a square of side "r" pixels from those
+ *        coordinates. In simpler terms, returns the automation node that was
+ *        clicked.
+ * @param Int X coordinate
+ * @param Int Y coordinate
+ * @param Boolean. True to check if the outValue of the node was clicked instead
+ *        (defaults to false)
+ * @param Int R distance in pixels
+ * @return timeMap::iterator with the clicked node, or timeMap.end() if none was clicked.
+ */
+AutomationEditor::timeMap::iterator AutomationEditor::getNodeAt(int x, int y, bool outValue /* = false */, int r /* = 5 */)
+{
+	QMutexLocker m(&m_patternEditorMutex);
+
+	// Remove the VALUES_WIDTH from the x position, so we have the actual viewport x
+	x -= VALUES_WIDTH;
+	// Convert the x position to the position in ticks
+	int posTicks = (x * MidiTime::ticksPerBar()/m_ppb) + m_currentPosition;
+
+	// Get our pattern timeMap and create a iterator so we can check the nodes
+	timeMap & tm = m_pattern->getTimeMap();
+	timeMap::iterator it = tm.begin();
+
+	// If there's a node that is inside those coordinates we will store it here
+	timeMap::iterator node = tm.end();
+
+	// ticksOffset is the number of ticks that match "r" pixels
+	int ticksOffset = MidiTime::ticksPerBar() * r / m_ppb;
+
+	while (it != tm.end())
```

Comment:
`it` is sorted by time, right? Then as soon as we're past the end of the range we can break out of the loop. Also, right now this appears to take the latest found point. If we're not taking the closest point, taking the first found one instead should be equally valid and exit the loop faster, right?

Reply:
I was improving this method and noticed that another thing can be further improved. `getNodeAt` gets called twice if we are in the DRAW_OUTVALUE mode, once for the inValue sphere and another for the outValue sphere (because the first call to `getNodeAt` is before the switch statement). Will fix that as well.